design/GraphCopyStabilization.md

```diff
@@ -0,0 +1,129 @@
+
+### Serialization
+* Cheney’s algorithm using main memory as from-space and stable memory as to-space: 
+* Focusing on stable variables as root (sub-graph of stable objects).
+* The target pointers and Cheney’s forwarding pointers denote the (skewed) offsets in stable memory.
+* Using streaming reads for the `scan`-pointer and streaming writes for the `free`-pointer in stable memory.
+
+### Deserialization
+* Cheney’s algorithm using stable memory as from-space and main memory as to-space: 
+* Starting with the stable root created during the serialization process.
+* A scan stack is used in the main memory to remember the deserialized objects for later scanning.
+* Objects are allocated in main memory using the default allocator.
+* Using random read/write access on the stable memory.
+
+## Stable Format
+For a long-term perspective, the object layout of the serialized data in the stable memory is fixed and independent of the main memory layout.
+* Pointers are represented in 64-bit like main memory in enhanced orthogonal persistence.
+* The Brooks forwarding pointer used by the incremental GC is omitted.
+* The pointers encode skewed stable memory offsets to the corresponding target objects.
+* References to the null objects are encoded by a defined null sentinel value.
+* `BigInt` are explicitly serialized in a defined portable little endian representation, without that the serialization or deserialization allocates temporary objects.
+The format is also versioned to allow future refinements of the graph copy algorithm.
+
+## Specific Aspects
+* Field hashes in objects are serialized in a blob. On deserialization, the hash blob is allocated in the dynamic heap. Same-typed objects that have been created by the same program version share the same hash blob.
+* Stable records can dynamically contain non-stable fields due to structural sub-typing. A dummy value can be serialized for such fields as a new program version can no longer access this field through the stable types.
+* For backwards compatibility, old Candid destabilzation is still supported when upgrading from a program that used older compiler version.
+* Incremental GC: Serialization needs to consider Brooks forwarding pointers (not to be confused with the Cheney's forwarding information), while deserialization can deal with partitioned heap that can have internal fragmentation (free space at partition ends).
+* The partitioned heap prevents linear scanning of the heap, especially in the presence of large objects that can be placed at a higher partition than subsequently allocated normal-sized objects. For this reason, a scan stack is allocated in the main memory, remembering the deserialized objects that still need to be scanned. With this, the deserialization does not need to make any assumptions of the heap structure (e.g. monotonically increasing allocations, free space markers, empty heap on deserialization start etc.).
```

Comment:
Do we need to worry about the scan stack overflowing at all? I assume not, but just checking..

Reply:
The scan stack grows in the heap. My assumption is that there is sufficient space in the heap. I can measure the maximum stack depth for some complex practical cases but I assume it is relatively low (maybe a few MB). Moreover, array slicing during the scanning (`scan_deserialized`) helps to reduce the stack depth for large arrays.

Reply:
(TODO for @luc-blaeser : Measure the scan stack depth.)

Reply:
I think you are probably fine since your stack is not fixed size but linked segments, right?

Reply:
It is only limited by heap free space...